gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/KillDagThread.java

```diff
@@ -0,0 +1,168 @@
+import org.apache.gobblin.service.monitoring.JobStatus;
+
+import static org.apache.gobblin.service.ExecutionStatus.CANCELLED;
+import static org.apache.gobblin.service.ExecutionStatus.ORCHESTRATED;
+import static org.apache.gobblin.service.ExecutionStatus.valueOf;
+
+
+@Slf4j
+public class KillDagThread implements Runnable {
+  private final Long defaultJobStartSlaTimeMillis;
+  NewDagManager dagManager;
+  public KillDagThread(Long defaultJobStartSlaTimeMillis, NewDagManager newDagManager) {
+    this.defaultJobStartSlaTimeMillis = defaultJobStartSlaTimeMillis;
+    this.dagManager = newDagManager;
+  }
+
+  @Override
+  public void run() {
+    for (Dag.DagNode<JobExecutionPlan> node : this.dagManager.getJobToDag().keySet()) {
+      boolean flowKilled = enforceFlowStartDeadline(node);
+      boolean jobKilled = false;
+
+      if (!flowKilled) {
+        JobStatus jobStatus = pollJobStatus(node);
+        try {
+          jobKilled = enforceJobCompletionDeadline(node, jobStatus);
+        } catch (ExecutionException | InterruptedException e) {
+          log.warn("Error getting status for dag node " + node.getId());
+          continue;
+        }
+      }
+
+      if (flowKilled || jobKilled) {
+        JobExecutionPlan jobExecutionPlan = DagManagerUtils.getJobExecutionPlan(node);
+        jobExecutionPlan.setExecutionStatus(CANCELLED);
+        try {
+          this.dagManager.onJobFinish(node);
+        } catch (IOException e) {
+          throw new RuntimeException(e);
+        }
+        String dagId = DagManagerUtils.generateDagId(node).toString();
+        this.dagManager.deleteJobState(dagId, node);
+      }
+    }
+  }
+
+  /**
+   * Retrieve the {@link JobStatus} from the {@link JobExecutionPlan}.
+   */
+  private JobStatus pollJobStatus(Dag.DagNode<JobExecutionPlan> dagNode) {
+    Config jobConfig = dagNode.getValue().getJobSpec().getConfig();
+    String flowGroup = jobConfig.getString(ConfigurationKeys.FLOW_GROUP_KEY);
+    String flowName = jobConfig.getString(ConfigurationKeys.FLOW_NAME_KEY);
+    long flowExecutionId = jobConfig.getLong(ConfigurationKeys.FLOW_EXECUTION_ID_KEY);
+    String jobGroup = jobConfig.getString(ConfigurationKeys.JOB_GROUP_KEY);
+    String jobName = jobConfig.getString(ConfigurationKeys.JOB_NAME_KEY);
+
+    return pollStatus(flowGroup, flowName, flowExecutionId, jobGroup, jobName);
+  }
+
+  private JobStatus pollStatus(String flowGroup, String flowName, long flowExecutionId, String jobGroup, String jobName) {
+    long pollStartTime = System.nanoTime();
+    Iterator<JobStatus> jobStatusIterator =
+        this.dagManager.getJobStatusRetriever().getJobStatusesForFlowExecution(flowName, flowGroup, flowExecutionId, jobName, jobGroup);
+    Instrumented.updateTimer(this.dagManager.getJobStatusPolledTimer(), System.nanoTime() - pollStartTime, TimeUnit.NANOSECONDS);
+
+    if (jobStatusIterator.hasNext()) {
+      return jobStatusIterator.next();
+    } else {
+      return null;
+    }
+  }
```

Comment:
this method seems like it would be used by other action threads too. can we put this into a util class?